include/experimental/__p1673_bits/conjugate_if_needed.hpp

```diff
@@ -0,0 +1,85 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 2.0
+//              Copyright (2019) Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software. //
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef LINALG_INCLUDE_EXPERIMENTAL___P1673_BITS_CONJUGATE_IF_NEEDED_HPP_
+#define LINALG_INCLUDE_EXPERIMENTAL___P1673_BITS_CONJUGATE_IF_NEEDED_HPP_
+
+#include <complex>
+
+namespace std {
+namespace experimental {
+inline namespace __p1673_version_0 {
+namespace linalg {
+namespace impl{
+
+template<class T> struct is_complex : std::false_type{};
+
+template<> struct is_complex<std::complex<float>> : std::true_type{};
+template<> struct is_complex<std::complex<const float>> : std::true_type{};
```

Comment:
Oof, I missed this bit.  I don't think you can instantiate `std::complex` with a `const`-qualified type.  Did you perhaps mean to write `is_complex<const std::complex<float>>`?

I think a better strategy would be to define `is_complex` to inherit from `true_type` only for `std::complex<float>`, `std::complex<double>`, and `std::complex<long double>`.  Then, use `std::remove_cv_t` on types before applying `is_complex` to them.

Reply:
Short answer: Only make `is_complex<T>` inherit from `true_type` for `T` = `std::complex<float>`, `std::complex<double>`, and `std::complex<long double>`.

Reply:
@mhoemmen 
`decay` is going to remove the `const` from something like `const T`, decaying it to just `T`. 
It seems to me that you can instantiate complex with const-qualified value_type, and if we don't provide the specializations for `is_complex` the way I did, it won't work, see this: https://godbolt.org/z/YzTG4ovYT

Reply:
@fnrizzi *[complex.numbers.general]* 2 says: "The effect of instantiating the template complex for any type other than `float`, `double`, or `long double` is unspecified."  We shouldn't support undefined behavior that happens to work in some implementations.

Reply:
thanks @mhoemmen for pointing this out! Sorry for dragging the discussion... I did not know about this detail ! 
Now I know.